### DISCUSSION TITLE
Improved ergonomics (Manual tool change, TLO procedure.)

### DISCUSSION THREAD

**Discussion by JipeB** [vlachoudis/bCNC#1827] [Ideas]
The erratic operation of this procedure is mainly due to the erroneous use of the G43.1 Zxxx function to compensate for the height of the measuring sensor when determining the WCS Z0 (see #1775)
 and to the ignorance of the management of the TLO by grbl.

By applying the correction proposed (see #1775) and the calibration procedure (https://github.com/vlachoudis/bCNC/wiki/Manual-tool-change-:-TLO-procédure), the manual tool change by the TLO procedure is perfectly operational.
This procedure has the advantage of not modifying the machining reference frame at each tool change and makes it possible to use the predetermined workspaces memorized by grbl.

A few modifications to improve ergonomics and simplify operation:

**CNC.py**
def toolChange()
*    simplification of the procedure
*    user choices (fastprobfeed, probfeed)
*   current tool display
*    current tool TLO display

**bmain.py**
def __init__(self, master, **kw):
* update and display TLO

**probePage.py**
defcalibrate()
*    simplification of the procedure
*    updated SensorHeight field

**probePage.py**
def updateTool(self):
*    New name to avoid confusion with tool number
*    Update display

def calibrate(self, event=None):
*    Improved readability of the procedure
*    Automatic update of sensorHeight at the end of the calibration

def _monitorSerial(self):
*   update (see #1775)

**controlPage.py**
def setTLO(self, event=None):
*   update display

JP
Details of changes to be made: [TLOmodif.pdf](https://github.com/vlachoudis/bCNC/files/11072134/TLOmodif.pdf)

**Comment by Harvie** [comment]
This sounds interresting,
However it's rather unconventional to share patch in PDF, can you please just prepare github pull request so we can review and merge it?

**Reply by JipeB** [reply]
Sorry, I'm not very familiar with these procedures; I will rephrase my edits.
JP

**Comment by Git-Bruno** [comment]
Hi JP
Please find below 2 fixes I did to get bCNC running with your changes. This is with python 3.8.10 & 3.10.11.
Fix 1.
Error message :
```
  File "C:\Users\Bruno\Documents\CFAO\CNC\bCNC-master_JipeB\bCNC-master\bCNC\CNC.py", line 750
    "sensor": ""
    ^
SyntaxError: invalid syntax
```
Change in CNC.py.
lines 749 & 750 to:
```
"Zsensor": 0.0,
"sensor": "",
```
Fix 2.
Error message :
```
  File "C:\Users\Bruno\Documents\CFAO\CNC\bCNC-master_JipeB\bCNC-master\bCNC\ProbePage.py", line 356, in __init__
    self.tlo.bind("<Return>", self.sensorSet)
AttributeError: 'NoneType' object has no attribute 'bind'
```
Change in ProbePage.py.
lines 356 & 357 to :
```
self.probeSensor.bind("<Return>", self.sensorSet)
self.probeSensor.bind("<KP_Enter>", self.sensorSet)
```

I cannot test on a machine now so I hope the code still works as you expect with the latter change.

**Comment by JipeB** [comment]
Hi Bruno,
Sorry, I've been using this code for two years on machines running Python 3.6.9 and 3.9.6 and must have made a few mistakes when I checked out my changes.
I haven't had a chance to get it to work with newer versions of Python.
Please keep me informed of your trials and thank you for the contact.
Best regards.

Also see : Jipe b patch 2 #1830
